### PR TITLE
ETS: Allow for conditional insertions

### PR DESCRIPTION
**Pull request by g-andrade**
This change introduces a new BIF, _ets:compare_insert/3_, for sets and ordered sets, which conditions object insertion based on what's in the table matching what we expect (in other words, a sort of compare-exchange for optimistic, atomic updates.)

As for the erl_db_util.h / DbTableMethod interface: I'm not sure whether having '_db_compare_put_' is better than simply extending the signature and functionality of '_db_put_', but that's fairly easy to change.

Supersedes #1074

### CODE REVIEW

**Comment by sverker** [comment]
The patch looks good.
There are however some resistance against new functions to the already fat ETS interface.
In this case I think it would be worth to at least consider a more versatile "select_replace" function with a match-spec as argument.

**Comment by g-andrade** [comment]
That's reasonable, and depending on implementation could make up for future specializations for elements, counters, etc. that would fatten the interface even more; I'll look into it.

**Comment by g-andrade** [comment]
OK, I've made a working draft for _db_hash_, based on the existing code for select_delete/2; what do you think? It returns the number of replaced objects and this is how it acts upon a match:
1. Is the match result a tuple? 
2. Does the match result have a key and is it the same as the matched object? 
3. If 1) and 2) apply, replace the object with the match result.

If you think this approach works, I'll wrap up the version for _db_tree_ and prepare some tests.

**Comment by OTP-Maintainer** [comment]
The summary line of the commit message is too long and/or ends with a "."
Make sure the whole message follows the guidelines here: https://github.com/erlang/otp/wiki/Writing-good-commit-messages.

Bad message: ETS: Allow for matchspec-based replacement on sets, bags and duplicate_bags


----------------

I am a script, I am not human 

----------------

**Comment by g-andrade** [comment]
Ok, I've dropped the 'ets:compare_insert/3' code and squashed the 'select_replace' implementation for _db_tree_ into the commit for _db_hash_.

**Comment by MSch** [comment]
Now that Erlang 19 is out, is this something that could make it into 19.1 or 19.2 or will this only be available once Erlang 20 is released?

**Comment by g-andrade** [comment]
(_db_tree_ (ordered_sets) is still not working properly - it crashes when the matchspec returns a differently sized object, as replacement then requires a reallocation and the parent node's pointer is not updated, leading to a segfault when traversing the tree; when I find the time I'll wrap up the fix)

**Comment by OTP-Maintainer** [comment]
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

**Comment by g-andrade** [comment]
All issues that popped up so far have been solved; I've added unit tests, which helped me track down a handful of problems as well.

So, this is the current status:
- select_replace/2 operation supported on all four table types
- the more objects we replace, the more reductions we bump (per trapped call)
- ordered_set peculiarities:
    - a matching key - not necessarily an equal key - is acceptable in the replacement object (e.g. 3.0 instead of 3), which is consistent with other operations
- bag peculiarities:
     - when we get a match, we exhaustively search the rest of the bucket for objects equal to the match result so that we don't end up with duplicate objects; this is very inefficient

I wonder if it could be more useful to return the actually replaced objects instead of a sum.

**Comment by g-andrade** [comment]
@MSch it would be really nice to have this for 19.x but I'm guessing it's too significant an interface change (assuming it's accepted.)

**Comment by vans163** [comment]
This will make it into mnesia module too?  Finally an atomic update vs the dreaded read+write transaction.

**Comment by OTP-Maintainer** [comment]
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

**Comment by sverker** [comment]
The `bag` case is a bit troublesome.
The current solution have cases when the outcome is dependent on the insertion order of the objects. For example to replace A -> B and B -> C will only succeed with both if A was inserted before B, that is if B->C is done before A->B. Not sure what the best solution is here.

Another problem is the complexity will be O(N^2) which can really hurt for large groups with same key.
One alternative is to disallow `select_replace` for `bag` at least until we have #420 with fast object lookup.

**Comment by g-andrade** [comment]
I've removed the `bag`-specific code and disallowed executing the operation over bags.

**Comment by g-andrade** [comment]
@vans163 seeing as ETS is the default backend for Mnesia, maybe this could form the basis for it, yes, but as we're then dealing with network, latency and other sorts of bottlenecks I wouldn't expect it to help much beyond syntactic sugar. 
Then again, I haven't looked at Mnesia for quite some time, so who knows?

**Comment by OTP-Maintainer** [comment]
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

**Comment by g-andrade** [comment]
Any news? Is disallowing the functionality for `bag` a reasonable option for it to go forward or will it wait on #420 for the time being?

**Comment by sverker** [comment]
Sorry for the silence.
I like this functionality. The restriction for bags is fine as a first step at least.

The only snag for me is the increased code bloat.
There are 6 similar variants of select iterations in erl_db_hash.c.
This PR adds 2 more. It would be really nice if that code could be refactored
into something more slim. Maybe something like in erl_db_tree.c
with a common iteration loop and different do_it-functions.

**Comment by g-andrade** [comment]
Well, that took me a while. So, here's the gist of what I've refactored the code into:
- Two generic iteration functions, `match_traverse` and `match_traverse_continue`, which receive callback pointers to specific "nothing can match", "handle match result", "loop ended" and "trap" stateful behaviours
- Common trapping / continuation unpacking code for `select_count`, `select_delete` and `select_replace`

There's still some duplicated code across `match_traverse` and `match_traverse_continue`, and the `select_chunk` callbacks ain't quite as simple as I would like, but I reckon it's better overall. What do you think?

**Comment by sverker** [comment]
I will hopefully get some time next week to have serious look.
After fixing one little typo when compiling `make TYPE=debug smp`
it went through the entire ets_SUITE. So far so good.

**Comment by g-andrade** [comment]
Some conflicts popped up in the meantime; I've taken the liberty of rebasing and (re)integrating these new 'magic ref' objects and everything seems to be ok.

**Comment by sverker** [comment]
I've started looking at the code. Kudos for doing that refactoring.

One thought on the semantics. Would it not be nicer behavior if select_replace started by verifying the matchspec to preserve the key and rejected the entire operation if it doesn't. I think it would work just by comparing the key position in the first match tuple with the corresponding position in the third build tuple and demand them to be equal. What do you think?

**Comment by g-andrade** [comment]
I'd say it would be a much more efficient way to verify the matchspec validity but it could lead to unexpected edge cases; e.g. let's say there's a match spec of this sort: 
`ets:fun2ms(fun (#tuple{ key = IntegerKey } = Object) -> Object#tuple{ key = IntegerKey + 0 } end)`
Which is perfectly safe but would get rejected under that check (contrived example, of course, but weird nonetheless.) On the other hand, the match spec could be explicitly rejected instead of simply having no effect, which is obviously better. I'm leaning towards the idea.

**Comment by g-andrade** [comment]
Another imagined scenario, perhaps a more realistic one, which would no longer be feasible - storing old object versions on objects themselves and using select_replace to restore previous versions:
`ets:fun2ms(fun (#foobar{ old_foo_version = OldObject }) -> OldObject end)`

**Comment by g-andrade** [comment]
I've though this over; perhaps it's best to blindly overwrite the key on the replacement tuple, while making it clear on the documentation that trying to modify it through select_replace won't have any effect. As we're already modifying / replacing the tuple, the relative overhead should be fairly low, especially on costlier replacements. 

However, if there's reason to worry about the computational cost, it could be a selective behaviour - by analyzing the matchspec it could be determined whether there's a risk of key replacement, and make it so that keys are overwritten on tuple replacement only if such a risk exists.

For example, this replacement matchspec is key-safe and no safeguard is needed:
`ets:fun2ms(fun (#person{ key = jane }  = Jane) -> Jane#person{ age = 42 } end)`

But this one might not be, and the key safeguard would have to be enforced:
`ets:fun2ms(fun (#person{ key = jane }) ->  ExternallyConstructedJane end)`

This safeguard could be enforced either on the traversal callback (more complex, more state to pass around) or by modifying the matchspec (simpler, a one-time check before compiling the match program.)

So, to summarize, I see four approaches to the problem:

1. Compare keys on every replacement, ignore replacement if they're not equal (db_hash) or if they don't match (db_tree) - current solution
2. Reject matchspecs that _might_ not be key safe, at the cost of not allowing operations like whole tuple replacements
3. Blindly overwrite the key on all replacement tuples
4. Selectively overwrite the key on replacement tuples if we can't be sure that the matchspec is key-safe

Which would you prefer?

**Comment by sverker** [comment]
To start with I think I prefer the conservative approach #2 and reject matchspecs that might not be key safe. It's always possible to add more features by allowing more advanced matchspecs if it should show to be very useful.

**Comment by g-andrade** [comment]
Unsafe match specs are now rejected right off the bat.

**Comment by sverker** [comment]
I feel like I tricked you into doing something less useful than I thought it would be.

The original use case of this PR, something like this with a constant key
`ets:replace(T, [{{Key, OldVal}, [], [{{Key, NewVal}}]}])`
used to work if key did not contain tuples, now it doesn't work at all.

I thought the check could be done by just comparing the two key terms for equality and cover both constants and variables. But the double tuple syntax for tuple construction complicates things.

Either go forward and write some `db_compare()` function that compensates for double tuples and maybe also the {const,Term} format.

Or revert and go for blind key overwrite. What do you think?

**Comment by g-andrade** [comment]
> I feel like I tricked you into doing something less useful than I thought it would be.

Oh well, I'll admit I also lost track of the intent behind the original feature... :-)

> I thought the check could be done by just comparing the two key terms for equality and cover both constants and variables. But the double tuple syntax for tuple construction complicates things.

Indeed, I now realize there's a few (more) corner cases `db_match_keeps_key` isn't accounting for.

> Or revert and go for blind key overwrite. What do you think?

I'd be happy with that.

**Comment by g-andrade** [comment]
Key-unsafe matchspecs are accepted again, and the keys for replacement objects are blindly overwritten.

**Comment by sverker** [comment]
To start from the end, I changed my mind (again) and I did a proposed improvement on your solution
to reject key-unsafe matchspecs here:
https://github.com/sverker/otp/commits/sverker/ets_conditional_insert/PR-1076

Rational:

- The blind overwrite implementation was not correct for non-immediate keys. A correct implementation would have to first copy the key out from the ets object before calling `replace_dbterm()` to then copy it back as part of the new ets object.

- Apart from the extra key-copy I realized that there is a temporary copy done for each replaced object. To optimize that (not necessary part of this PR), and let the matchspec machine allocate and build the ets object, will be simpler (or even not impossible) if the key is known to be correct when the term is built.

- I like the clean semantics of letting _only_ the match and guard parts determine if the third body part should be done or not.

I'm committed to this PR and think `ets:select_replace` should be included in OTP-20.
This is what's need to be done before merge to master (regardless who is doing it):

- Rebase on master. We have a rather big job going on changing table identifiers to be references. We have not yet merged this job to master. It will not conflict much with this PR but some manual work will be needed in erl_db_hash.c I think.

- Try remove the garbage. I think it would be worth at least try change so the matchspec machine
builds on a temporary heap to avoid filling up the heap of the calling process leading to unnecessary garbage collections. There is a flag called ERTS_PAM_TMP_RESULT used when running tracing matchspec that does that, but I'm not sure it's just that easy to change it from ERTS_PAM_COPY_RESULT.

/Sverker

**Comment by g-andrade** [comment]
> The blind overwrite implementation was not correct for non-immediate keys. A correct implementation would have to first copy the key out from the ets object before calling replace_dbterm() to then copy it back as part of the new ets object.

Ahh, indeed, I missed those.

> Try remove the garbage. I think it would be worth at least try change so the matchspec machine
builds on a temporary heap to avoid filling up the heap of the calling process leading to unnecessary garbage collections. There is a flag called ERTS_PAM_TMP_RESULT used when running tracing matchspec that does that, but I'm not sure it's just that easy to change it from ERTS_PAM_COPY_RESULT.

I've started looking into it and it doesn't appear to be a trivial change; switching the flag to ...TMP_RESULT and getting rid of no longer necessary process heap management steps resulted in crashes when attempting to store the object, which I eventually traced back to the fact that the match spec had been compiled with DCOMP_TABLE (I think.) Compiling it with DCOMP_TRACE no longer induces crashes but gives out wrong results, so I'll still have to figure out how do these DCOMP flags exactly influence the behavior of the match program.

Everything I've just said might be wrong, as I'm not (yet) entirely aware of what I'm doing :-)

**Comment by sverker** [comment]
I think I solved the "Try remove the garbage" issue.

I pushed three new commit to
https://github.com/sverker/otp/commits/sverker/ets_conditional_insert/PR-1076

**Comment by g-andrade** [comment]
Indeed! It's working well; the solution was quite simple after all. It would have taken me a while to figure out why the match machine stack didn't appear to contain what I was expecting - not all the time, anyway, and assuming I'm interpreting ea12abf correctly. I've pushed your fixes.

So I suppose the only thing remaining now is waiting for the upcoming table identifier changes to be merged to master and solve any conflicts that arise.

**Comment by sverker** [comment]
Yes, I think it's ready for rebase/merge now.

But in the meantime, if I'm allowed to nitpick and you have some time to spend on polish...

- State outright in docs that matchspec must retain key.
- Make clearer in docs that the atomicity of select_replace only applies per object and not over the iteration (as all other select variants).
- I think the bag-warning is a bit much. Isn't it enough if that is just mentioned in the text.
- I don't like local variable initializations that are never used.
  I think it makes the code harder to understand and  gcc is pretty good at detecting use of uninitialized variables and our valgrind runs are even better.

**Comment by sverker** [comment]
We have now merged the branch that changed table id from integers to refs
here https://github.com/erlang/otp/commit/b4ac8b2b32f094217d0533ee139273923c3a8af7

The commit that will cause rebase conflicts is https://github.com/erlang/otp/commit/e214bb4a0f86a6f956dd473f66b1a6c2f7085591 that replaces `tb->common.id` with a `tid` argument that is passed down to some of the select functions.

I will not do this rebase right now, so if you feel the urge...

**Comment by g-andrade** [comment]
I've rebased over master and integrated the new table reference ids in both `select_replace` implementations (tree and hash.) I've also polished the documentation a bit and removed any redundant variable initializations that caught my eye.

**Comment by sverker** [comment]
Great! I will put it back into our test runs for master.

**Comment by sverker** [comment]
Mnesia tests found a bug that ets_SUITE did not cover.

I pushed a fix to YOUR https://github.com/g-andrade/otp/tree/feature/ets_conditional_insert

**Comment by sverker** [comment]
I fixed some use of uninitialized values found by valgrind.
